### PR TITLE
feat: Properly demangle Swift class name

### LINKED ISSUE #1980: Decode Swift type names

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Add Swift code into the framework and use it to properly demangle Swift class names.

## :bulb: Motivation and Context

see https://github.com/getsentry/sentry-cocoa/discussions/2160

closes #1980

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [ ] No breaking changes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -62,8 +62,8 @@ build-xcframework-sample:
 # Building the .frameworsk.zip only works with Xcode 12, as there is no workaround yet for Xcode 13.
 build-framework:
 	@echo "--> Carthage: creating Sentry framework"
-	./scripts/carthage-xcode12-workaround.sh build --no-skip-current
-	./scripts/carthage-xcode12-workaround.sh archive Sentry --output Sentry.framework.zip
+	./scripts/carthage-xcode12-workaround.sh build --no-skip-current --use-xcframeworks
+	./scripts/carthage-xcode12-workaround.sh archive Sentry --use-xcframeworks --output Sentry.framework.zip
```
Comment (philipphofmann) [inline Makefile:66]: `h`: This step should not use xcframeworks, why did you add this? It should create a `.framework.zip` and not a `.xcframework.zip`.

**Comment by brustolin** [comment]
Projects using Carthage will be required to target both Sentry and SentryObjc frameworks.

**Comment by brustolin** [comment]
@philipphofmann @andrewlvovsky @kahest 
This PR is ready for review. The changes here are most related to the Workspace configuration to support swift in the project

File: Sources/Sentry/include/SentrySwift.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#ifndef SentrySwift_h
+#define SentrySwift_h
+
+#if SWIFT_PACKAGE
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySwift.h:4]: `m`: Please add a comment on why we need this.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 ### Features
 
+- Sentry Swift framework (#2162)
```
Comment (philipphofmann) [inline CHANGELOG.md:7]: `m`: I think this PR fixes https://github.com/getsentry/sentry-cocoa/issues/1980, which is the only new feature in this PR. Therefore I would adapt the name to reflect this change.

Comment (philipphofmann) [inline CHANGELOG.md:7]: `h`: We need to point out here and also in the docs that from now on, the SDK has a dependency on Swift.

File: Sentry.podspec [hunk 4]
```diff
@@ -16,6 +16,7 @@ Pod::Spec.new do |s|
   s.requires_arc = true
   s.frameworks = 'Foundation'
   s.libraries = 'z', 'c++'
+  s.swift_versions = "5.0"
```
Comment (philipphofmann) [inline Sentry.podspec:19]: `h`: Can't we already choose a higher Swift version?

File: Sources/Sentry/Public/SentryStacktrace.h [hunk 5]
```diff
@@ -7,7 +7,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 @class SentryFrame;
 
-NS_SWIFT_NAME(Stacktrace)
```
Comment (philipphofmann) [inline Sources/Sentry/Public/SentryStacktrace.h:10]: `h`: This is a breaking change, and we must wait until the next major to do this.

File: Sources/Sentry/Public/SentryThread.h [hunk 6]
```diff
@@ -5,7 +5,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 @class SentryStacktrace;
 
-NS_SWIFT_NAME(Thread)
```
Comment (philipphofmann) [inline Sources/Sentry/Public/SentryThread.h:8]: `h`: This is a breaking change, and we must wait until the next major to do this.

File: Sources/Sentry/SentryUIViewControllerSanitizer.m [hunk 7]
```diff
@@ -1,34 +1,11 @@
 #import "SentryUIViewControllerSanitizer.h"
+#import "SentrySwift.h"
 
 @implementation SentryUIViewControllerSanitizer
```
Comment (philipphofmann) [inline Sources/Sentry/SentryUIViewControllerSanitizer.m:4]: `m`: I think we can ditch this class now. I only calls `SwiftDescriptor getDescription`.

File: Package.swift [hunk 8]
```diff
@@ -9,8 +9,15 @@ let package = Package(
         .library(name: "Sentry-Dynamic", type: .dynamic, targets: ["Sentry"])
     ],
     targets: [
+        .target( name: "SentrySwift",
+                 path: "Sources",
+                 sources: [
+                    "SentrySwift"
```
Comment (philipphofmann) [inline Package.swift:15]: `m`: I think we don't need to prefix that. We could rename the folder to `Swift` only. What do you think?
```suggestion
                    "Swift"
```

File: Sources/SentrySwift/SwiftDescriptor.swift [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+@objc
+public class SwiftDescriptor: NSObject {
+ 
+    @objc
+    public static func getDescription(_ object: AnyObject) -> String {
```
Comment (philipphofmann) [inline Sources/SentrySwift/SwiftDescriptor.swift:7]: `h`: I think this must not be public.
```suggestion
class SwiftDescriptor: NSObject {
 
    @objc
    static func getDescription(_ object: AnyObject) -> String {
```

**Review by philipphofmann — COMMENTED** [review]
It's great to see this PR landing finally, and I'm happy that we only need one target. Please add an entry to the [decision log of the develop docs](https://github.com/getsentry/sentry-cocoa/tree/master/develop-docs#decision-log) on why we decided to add a dependency to Swift.

File: Sources/SentrySwift/SwiftDescriptor.swift (hunk 9 above)
Comment (brustolin) [reply Sources/SentrySwift/SwiftDescriptor.swift:7]: This need to be public in order to Objc to use it.
Swift code belongs in a different module, therefore it needs to be public.

But, to use this in a project, you need to import SentrySwift

Comment (philipphofmann) [reply Sources/SentrySwift/SwiftDescriptor.swift:7]: I removed the `public`, and it compiled. Does the compilation then only fail for the package managers? This class shouldn't be public. Nobody has to use this.

> But, to use this in a project, you need to import SentrySwift

So if we add actual public Swift code now, the user needs to add a different import? I think that is not good. I would like to avoid that.

Comment (brustolin) [reply Sources/SentrySwift/SwiftDescriptor.swift:7]: SPM fails if you remove the public.

> So if we add actual public Swift code now, the user needs to add a different import? I think that is not good. I would like to avoid that.

Thats why Swift will be only available to help us. But we can't create public api in Swift. I think we could make swift code available through `#import <Sentry/Sentry.h>` but then everything in swift will be public.

Comment (brustolin) [reply Sources/SentrySwift/SwiftDescriptor.swift:7]: Hummm, For cocoa and Carthage everything in Swift will be public already 😕 for this approach

File: Sentry.podspec (hunk 4 above)
Comment (brustolin) [reply Sentry.podspec:19]: Sure, I dont see why not.

File: Sentry.podspec [hunk 10]
```diff
@@ -16,6 +16,7 @@ Pod::Spec.new do |s|
   s.requires_arc = true
   s.frameworks = 'Foundation'
   s.libraries = 'z', 'c++'
+  s.swift_versions = "5.5"
```
Comment (philipphofmann) [inline Sentry.podspec:19]: `h`: I think we should change this to
```suggestion
  s.swift_versions = ['5']
```

According to the [podspec](https://guides.cocoapods.org/syntax/podspec.html)

> The versions of Swift that the specification supports. A version of '4' will be treated as '4.0' by CocoaPods and not '4.1' or '4.2'.
> 
> Note The Swift compiler mostly accepts major versions and sometimes will honor minor versions. While CocoaPods allows specifying a minor or patch version it might not be honored fully by the Swift compiler.

So it mostly accepts major versions anyways, and why do we already want to jump to 5.5? Can't we still support 5.0?

Maybe 5.5 because it still works with Xcode 13.0, which still allows developing for iOS 9, which is our min iOS version? If yes, please put this info as a comment here.

File: Sources/Sentry/include/SentrySwift.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#ifndef SentrySwift_h
+#define SentrySwift_h
+
+/*
+ * This is a header to expose Swift code to Objective-c
+ *
+ * Because we use three different package managers
+ * we need to target the swift module in different ways.
+ *
+ * If the project is being used through SPM, Swift and Objective-c are compiled into separated
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySwift.h:10]:
```suggestion
 * If the project is being used through SPM, Swift and Objective-C are compiled into separated
```

File: Sources/Sentry/include/SentrySwift.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * targets, then we use "@import SentrySwift".
+ *
+ * Cocoapods combines everything into one target, if the user enable USE_FRAMEWORKS,
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySwift.h:13]:
```suggestion
 * CocoaPods combines everything into one target, if the user enables USE_FRAMEWORKS,
```

File: Sources/Sentry/include/SentrySwift.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
+#ifndef SentrySwift_h
+#define SentrySwift_h
+
+/*
+ * This is a header to expose Swift code to Objective-c
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySwift.h:5]: Thank you for writing a comment that describes why we need this.

File: develop-docs/README.md [hunk 14]
```diff
@@ -128,3 +128,14 @@ iOS 12 opens a risk of introducing bugs, which has already happened in the past,
 the iOS 12 simulator a try.
 
 Related to [GH-2218](https://github.com/getsentry/sentry-cocoa/issues/2218)
+
+### Adding Swift code in the project
+
+Date: October 1st 2022
+Contributors: @brustolin
+
+A Sentry SDK started to be [written in Swift once,](https://github.com/getsentry/raven-swift) but due to ABI not being stable at that time, it got dropped. Since then Swift 5.1 landed and we got ABI stability. We’ve considered adding Swift to our sentry.cocoa SDK since then, but because of some of the trade offs, we’ve postponed that decision.
```
Comment (philipphofmann) [inline develop-docs/README.md:137]: `l`: Swift 5.0 brought ABI stability, see https://www.swift.org/blog/swift-5-released/
```suggestion
A Sentry SDK started to be [written in Swift once,](https://github.com/getsentry/raven-swift) but due to ABI not being stable at that time, it got dropped. Since then Swift 5.0 landed and we got ABI stability. We’ve considered adding Swift to our sentry.cocoa SDK since then, but because of some of the trade offs, we’ve postponed that decision.
```

File: develop-docs/README.md [hunk 15]
```diff
@@ -128,3 +128,14 @@ iOS 12 opens a risk of introducing bugs, which has already happened in the past,
...
+This changed with our goal to better support SwiftUI. It’s growing in popularity and we need to write code in Swift in order to support it.
+SwiftUI support will be available through an adicional library, but in order to support it we need to be able to demangle Swift class names in Sentry SDK, which can be done by using Swift API.
```
Comment (philipphofmann) [inline develop-docs/README.md:139]: `l`: Typos
```suggestion
SwiftUI support will be available through an additional library, but to support it, we need to be able to demangle Swift class names in Sentry SDK, which can be done by using Swift API.
```

File: develop-docs/README.md [hunk 16]
```diff
@@ -128,3 +128,14 @@ iOS 12 opens a risk of introducing bugs, which has already happened in the past,
...
+Since we support SPM, and SPM don't support multi language projects, we need to create two different targets, one with Swift and another with Objective-c code, because of that our Swift code in SPM resides in a different module, which requires a different import in projects, and for now, this is the reason we should avoid writing public API in Swift. But, everything written in Swift will be public if users import Sentry via CocoaPods or Carthage.
```
Comment (philipphofmann) [inline develop-docs/README.md:140]: `h`: Can't we remove that?

> But everything written in Swift will be public if users import Sentry via CocoaPods or Carthage.

**Review by philipphofmann — COMMENTED** [review]
Do we still have to make the Swift code public? Or am I missing something?

File: Sources/SentrySwift/SwiftDescriptor.swift (hunk 9 above)
Comment (philipphofmann) [reply Sources/SentrySwift/SwiftDescriptor.swift]: As discussed in the stand-up, I think we should try to avoid this.

File: develop-docs/README.md [hunk 17]
```diff
@@ -128,3 +128,14 @@ iOS 12 opens a risk of introducing bugs, which has already happened in the past,
 the iOS 12 simulator a try.
 
 Related to [GH-2218](https://github.com/getsentry/sentry-cocoa/issues/2218)
+
+### Adding Swift code in the project
```
Comment (philipphofmann) [inline develop-docs/README.md:132]: Thank you, @brustolin, for adding an entry here, so we know in the future why we decided that way.

File: CHANGELOG.md [hunk 18]
```diff
@@ -4,6 +4,8 @@
 
 ### Features
 
+- Properly demangle Swift class name (#2162)
+  - From now on, Sentry SDK has a dependency on Swift framework
```
[the remark this answers is not in the document]
Comment (brustolin) [reply CHANGELOG.md:8]: > I would put this directly below the release version

How this should be? I haven't found an example of adding extra information to a release name. In the same line after the version number looks ugly.

File: Sentry.podspec (hunk 10 above)
Comment (brustolin) [reply Sentry.podspec:19]: It was your idea to support a version bigger than 5, thats why I changed. It made sense to me. 
Our CocoaPods supports 5.4, since this will be a major, why not ditch one minor. 5.5 is supported by Xcode 12, and this is the minimum Xcode we test anyway.

Comment (philipphofmann) [reply Sentry.podspec:19]: Could setting this to `5.5` break someone?

File: develop-docs/README.md (hunk 16 above)
Comment (philipphofmann) [reply develop-docs/README.md]: > We need to be very aware of everything we write in Swift right now because of this last sentence.

I thought we found a way to avoid everything being public. As discussed in the stand-up, we should avoid this.

File: Sentry.podspec (hunk 10 above)
Comment (brustolin) [reply Sentry.podspec:19]: Only if Xcode don't support 5.5. 
Which means there is someone out there using Xcode 11. 
I don't believe this is a thing.

File: CHANGELOG.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (brustolin) [reply CHANGELOG.md:8]: I see. This is not a PR information, this is a Version information, totally make sense. Thanks

File: Sources/SentrySwift/SwiftDescriptor.swift (hunk 9 above)
Comment (brustolin) [reply Sources/SentrySwift/SwiftDescriptor.swift]: The solution will be to remove this Class from Sentry module, then, when a user import Sentry, this will not show in the auto complete. 

It's not possible to remove the public from this class and functions.

Comment (philipphofmann) [reply Sources/SentrySwift/SwiftDescriptor.swift]: @brustolin, this is the only issue stopping me from approving the PR. Did you already find a solution to it? How can I support you?

**Comment by brustolin** [comment]
We decide that we're fine having swift class public in a different module, called SentryPrivate.
For this to happen we need an extra cocoapod library. SPM and Carthage will work like this out of the box.

With this approach, classes from SentryPrivate will not be available when users import Sentry.
We don't mind breaking changes in SentryPrivate, because this is not meant to be use by the user, we going to point this out in the docs.

Participants: @philipphofmann @kahest @brustolin

File: develop-docs/README.md [hunk 19]
```diff
@@ -135,3 +135,13 @@ iOS 12 opens a risk of introducing bugs, which has already happened in the past,
...
+
+Date: October 1st 2022
+Contributors: @brustolin
+
+A Sentry SDK started to be [written in Swift once,](https://github.com/getsentry/raven-swift) but due to ABI not being stable at that time, it got dropped. Since then Swift 5.0 landed and we got ABI stability. We’ve considered adding Swift to our sentry.cocoa SDK since then, but because of some of the trade offs, we’ve postponed that decision.
+This changed with our goal to better support SwiftUI. It’s growing in popularity and we need to write code in Swift in order to support it.
+SwiftUI support will be available through an additional library, but to support it, we need to be able to demangle Swift class names in Sentry SDK, which can be done by using Swift API.
+Since we support SPM, and SPM doesn't support multi-language projects, we need to create two different targets, one with Swift and another with Objective-C code. Because of that, our Swift code in SPM resides in a different module, which requires a different import in projects, and for now, this is the reason we should avoid writing public API in Swift.
```
Comment (philipphofmann) [inline develop-docs/README.md:147]: `m`: I think that part here needs updating with the information from the discussion we had today

File: Sources/Sentry/include/SentrySwift.h [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+ * Because we use three different package managers
+ * we need to target the Swift module in different ways.
+ *
+ * If the project is being used through SPM, Swift and Objective-C are compiled into separated
+ * targets, then we use "@import SentrySwift".
+ *
+ * CocoaPods combines everything into one target, if the user enables USE_FRAMEWORKS,
+ * Swift will be available through "#import <Sentry/Sentry-Swift.h>" otherwise "#import
+ * "Sentry-Swift.h"".
+ */
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentrySwift.h:16]: `m`: Is this comment still valid?

File: Sources/Swift/Sentry.swift [hunk 21]
```diff
@@ -0,0 +1,3 @@
+//This file is only necessary to make Sentry framework compile using swift compiler.
```
Comment (philipphofmann) [inline Sources/Swift/Sentry.swift:1]: `m`: Please add the reason why.

**Review by philipphofmann — COMMENTED** [review]
I think we should add a section, like `Code Signing` to the [cocoa develop docs](https://github.com/getsentry/sentry-cocoa/tree/master/develop-docs) to explain what how the package managers are set up and what SentryPrivate is.

File: Sources/Sentry/include/SentrySwift.h (hunk 20 above)
Comment (brustolin) [reply Sources/Sentry/include/SentrySwift.h:16]: nope

Comment (brustolin) [inline Sources/Sentry/include/SentrySwift.h:16]:
```suggestion
```

File: .github/workflows/lint.yml [hunk 22]
```diff
@@ -39,8 +39,8 @@ jobs:
       - run: ./scripts/ci-select-xcode.sh
       - run: make analyze
 
-  validate-podspec:
-    name: Validate Podspec
+  validate-podspec-framework:
```
Comment (philipphofmann) [inline .github/workflows/lint.yml:42]: Makes sense thanks 🙏

File: SentryPrivate.podspec [hunk 23]
```diff
@@ -0,0 +1,28 @@
+Pod::Spec.new do |s|
+  s.name         = "SentryPrivate"
+  s.version      = "7.30.2"
```
Comment (philipphofmann) [inline SentryPrivate.podspec:3]: `m`: 
```suggestion
  s.version      = "7.31.0"
```

File: develop-docs/README.md [hunk 24]
```diff
@@ -136,6 +136,19 @@ the iOS 12 simulator a try.
...
+
+Date: October 1st 2022
+Contributors: @brustolin
+
+A Sentry SDK started to be [written in Swift once,](https://github.com/getsentry/raven-swift) but due to ABI not being stable at that time, it got dropped. Since then Swift 5.0 landed and we got ABI stability. We’ve considered adding Swift to our sentry.cocoa SDK since then, but because of some of the trade offs, we’ve postponed that decision.
+This changed with our goal to better support SwiftUI. It’s growing in popularity and we need to write code in Swift in order to support it.
+SwiftUI support will be available through an additional library, but to support it, we need to be able to demangle Swift class names in Sentry SDK, which can be done by using Swift API.
+Since we support SPM, and SPM doesn't support multi-language projects, we need to create two different targets, one with Swift and another with Objective-C code. Because of that, our swift code needs to be public, so we're creating a second module called SentryPrivate, where all swift code will be, and we need an extra cocoapod library. 
+With this approach, classes from SentryPrivate will not be available when users import Sentry.
```
Comment (philipphofmann) [inline develop-docs/README.md:148]: 💯 Thanks Dhiogo

**Review by philipphofmann — COMMENTED** [review]
I'm a bit afraid of breaking the Hybrid SDKs again. How can we ensure that this is not happening?

File: Sentry.podspec [hunk 25]
```diff
@@ -26,20 +27,20 @@ Pod::Spec.new do |s|
   }
 
   s.default_subspecs = ['Core']
+  s.dependency "SentryPrivate", "7.30.2"
```
Comment (brustolin) [inline Sentry.podspec:30]:
```suggestion
  s.dependency "SentryPrivate", "7.31.0"
```

**Comment by brustolin** [comment]
We have two options. 
- Coordinate some tests across SDKs
- Create an integration CI test in our repo that downloads hybrids SDK sample and try to run it.

The second one will be kind heavy thing to add into our daily process.

**Review by philipphofmann — APPROVED** [review]
Let's do a beta release to mitigate the risk of breaking people. Before you merge, I think we should first coordinate with Hybrid SDKs so they are aware they need to test this. LGTM.

**Comment by philipphofmann** [comment]
Let's merge this to the 8.0.0 branch once we create it. I still would like to release a hotfix with https://github.com/getsentry/sentry-cocoa/pull/2398 before that.
